src/mlpack/methods/reinforcement_learning/ppo_impl.hpp

```diff
@@ -0,0 +1,189 @@
+/**
+ * @file ppo.hpp
+ * @author Xiaohong Ji
+ *
+ * This file is the implementation of PPO class, which implements
+ * proximal policy optimization algorithms.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_PPO_IMPL_HPP
+#define MLPACK_METHODS_RL_PPO_IMPL_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include "ppo.hpp"
+
+namespace mlpack {
+namespace rl {
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::PPO(TrainingConfig config,
+       ActorNetworkType ActorNetwork,
+       CriticNetworkType CriticNetwork,
+       PolicyType policy,
+       ReplayType replayMethod,
+       UpdaterType updater,
+       EnvironmentType environment):
+  config(std::move(config)),
+  ActorNetwork(std::move(ActorNetwork)),
+  CriticNetwork(std::move(CriticNetwork)),
+  updater(std::move(updater)),
+  policy(std::move(policy)),
+  replayMethod(std::move(replayMethod)),
+  environment(std::move(environment))
+{
+  if (ActorNetwork.Parameters().is_empty())
+    ActorNetwork.ResetParameters();
+  if (CriticNetwork.Parameters().is_empty())
+    CriticNetwork.ResetParameters();
+
+  this->updater.Initialize(ActorNetwork.Parameters().n_rows,
+                           ActorNetwork.Parameters().n_cols);
+  this->OldActorNetwork = ActorNetwork;
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+void PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::Update()
+{
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+double PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::Step()
+{
+  // Get the action value for each action at current state.
+  arma::colvec actionValue;
+  ActorNetwork.Predict(state.Encode(), actionValue);
```

Comment:
Here actionValue can't be a `colvec`, `ActorNetwork` expects `arma::mat` type to be pass as output, see https://github.com/mlpack/mlpack/blob/6bed414710b332549e4a44915d6dc4e7cc52917e/src/mlpack/methods/ann/ffn.hpp#L138

So, in this case `actionValue.col(0)` will probably be used for mean and rest of the rows will be used for variance. Hence, you can easily extract mean as a `vec` from `actionValue` and variance as a `mat` by deleting first row of `actionValue`.
Does this clear things up?

Reply:
> Here actionValue can't be a `colvec`, `ActorNetwork` expects `arma::mat` type to be pass as output, see
> 
> https://github.com/mlpack/mlpack/blob/6bed414710b332549e4a44915d6dc4e7cc52917e/src/mlpack/methods/ann/ffn.hpp#L138
> 
> So, in this case `actionValue.col(0)` will probably be used for mean and rest of the rows will be used for variance. Hence, you can easily extract mean as a `vec` from `actionValue` and variance as a `mat` by deleting first row of `actionValue`.
> Does this clear things up?

Okay, why I use `colvec` here because the code in [`q_learning.hpp` ](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/reinforcement_learning/q_learning_impl.hpp#L99), so I am not sure. In this situation, `mat` is more suitable.

Reply:
Ohh you are right. Basically the thing is, the functions that expects `mat` as an input can generally also accept `col`. So, here you can just grab values as `colvec` by the logic I mentioned earlier and should be able to pass it to the GaussianDistribution constructor, even though it expects a `mat` type to passed.

Reply:
If they don't accept the input, you can also checkout http://arma.sourceforge.net/docs.html#conv_to

Reply:
Hi, I have confused with the `EnvironmentType::Action`, why it can convert to `int` to the `EnvironmentType::Action` when use`CartPole`' in [line](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/reinforcement_learning/policy/greedy_policy.hpp#L70).

But when I use the code  to sample action for `Pendulum` task like this:
```
ActionType action = normalDist.Random();
```
the complier output like that:
```
: error: 
      no matching conversion for static_cast from 'arma::vec' (aka 'Col<double>') to 'ActionType' (aka
      'mlpack::rl::Pendulum::Action')
  ActionType action = static_cast<ActionType>(normalDist.Random());
```

Reply:
The ``Pendulum`` env does use an array as action, https://github.com/mlpack/mlpack/blob/e2f696cfd5b7ccda2d3af1c7c728483ea6591718/src/mlpack/methods/reinforcement_learning/environment/pendulum.hpp#L85

you can remove the ``[1]`` here and just use action everywhere else to solve the issue. But we should open a PR to fix this in the codebase.

Reply:
Hopefully that makes sense, let me know if I should clarify anything.

Reply:
> Hopefully that makes sense, let me know if I should clarify anything.

Sorry for late replay, I still struggle for this part. I remove the `[1]` everywhere, and remain the sample action code.
```
ActionType action = normalDist.Random();
```

It still the same error. 

```
: error: 
      no viable conversion from 'arma::vec' (aka 'Col<double>') to 'ActionType' (aka
      'mlpack::rl::Pendulum::Action')
  ActionType action = normalDist.Random();
             ^        ~~~~~~~~~~~~~~~~~~~
```

I am trying to figure out what's happening.

Reply:
> > Hopefully that makes sense, let me know if I should clarify anything.
> 
> Sorry for late replay, I still struggle for this part. I remove the `[1]` everywhere, and remain the sample action code.
> 
> ```
> ActionType action = normalDist.Random();
> ```
> 
> It still the same error.
> 
> ```
> : error: 
>       no viable conversion from 'arma::vec' (aka 'Col<double>') to 'ActionType' (aka
>       'mlpack::rl::Pendulum::Action')
>   ActionType action = normalDist.Random();
>              ^        ~~~~~~~~~~~~~~~~~~~
> ```
> 
> I am trying to figure out what's happening.

I rewrite the code in this way, it works but not so elegant and extensible.  

```
  ActionType action;
  action.action = normalDist.Random()[0];
```

Reply:
> your best bet

 @manish7294, thank you for your explanation.